File: Lab5_hank/tutorial/compiler.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#define ERR(error) perror(error); exit(0);

FILE *in_file;
FILE * out_file;
char dec_out[8];
char reg_out[8];
int counter;


void appendchar(const char * dest, int c)
{
    size_t len = strlen(dest);
    char *str2 = malloc(len + 1 + 1 ); /* one for extra char, one for trailing zero */
    strcpy(str2, dest);
    str2[len] = c;
    str2[len + 1] = '\0';
    free( str2 );
}
void dec2bin(int val){
    int rem;
    int next = val;
    char tmp;

    strcpy(dec_out,00000000);
    while (next != 0){
        rem = next%2;
        tmp = rem+'0';
        strcat(dec_out,&tmp);
        next /= 2;
    }
}

void reg2bin(const char * reg){
    if (strcmp(reg,"R0"))     {strcpy(reg_out,"00");}
    else if(strcmp(reg,"R1")) {strcpy(reg_out,"01");}
    else if(strcmp(reg,"R2")) {strcpy(reg_out,"10");}
    else if(strcmp(reg,"R3")) {strcpy(reg_out,"11");}
}

void main(void){
    char operand[10];
    char str1[10],str2[10];
    char output[10];
    char fileName[64];
    int int1;
    char tmp;
    
    
    printf("file name: 3tlab5.txt\n");
    // strcpy(fileName,"3tlab5.txt");
    // printf("file name: ");
    // if (scanf("%s",fileName)<=0){ERR("scanf");}

    in_file = fopen("3tlab5.txt","r");
    out_file = fopen("outfile.txt","w");

    counter = 0;
    printf("counter = %d\n",counter);
    while((tmp = fgetc(in_file))!=EOF){
        printf("tmp = %c\n",tmp);
        if((tmp == ' ')|| (tmp == ',')) {counter =1;printf("counter = %d\n",counter);}
        
        if(counter <= 0){appendchar(operand,tmp);printf("operand = %s\n",operand);}
        else if (counter <= 1)
        {
            printf("operand = %s\n",operand);
            if(strcmp(operand,"BR")==0)
            {
                strcpy(output,"100");
                fscanf(in_file,"%d\n",int1);
                printf("int1 = %d\n",int1);
                if(int1 <= 1){strcat(output,"0000");}
                else if(int1 <= 3){strcat(output,"000");}
                else if(int1 <= 7){strcat(output,"00");}
                else if(int1 <= 15){strcat(output,"0");}
                dec2bin(int1);
                strcat(output,dec_out);
            }
            else if(strcmp(operand,"BRZ")==0)
            {
                strcpy(output,"101");
                fscanf(in_file,"%d\n",int1);
                printf("int1 = %d\n",int1);
                if(int1 <= 1){strcat(output,"0000");}
                else if(int1 <= 3){strcat(output,"000");}
                else if(int1 <= 7){strcat(output,"00");}
                else if(int1 <= 15){strcat(output,"0");}
                dec2bin(int1);
                strcat(output,dec_out);
            } 
            else if(strcmp(operand,"ADDI")==0)
            {
                strcpy(output,"000");
                fscanf(in_file,"%s, %d\n",str1,int1);
                printf("str1 = %s\n",str1);
                printf("int1 = %d\n",int1);
                if(int1 <= 1){strcat(output,"00");}
                else if(int1 <= 3){strcat(output,"0");}
                dec2bin(int1);
                strcat(output,dec_out);
                reg2bin(str1);
                strcat(output,reg_out);
            }
            else if(strcmp(operand,"SUBI")==0)
            {
                strcpy(output,"001");
                fscanf(in_file,"%s, %d\n",str1,int1);
                printf("str1 = %s\n",str1);
                printf("int1 = %d\n",int1);
                if(int1 <= 1){strcat(output,"00");}
                else if(int1 <= 3){strcat(output,"0");}
                dec2bin(int1);
                strcat(output,dec_out);
                reg2bin(str1);
                strcat(output,reg_out);
            }
            else if(strcmp(operand,"SR0")==0)
            {
                strcpy(output,"0100");
                fscanf(in_file,"%d\n",int1);
                printf("int1 = %d\n",int1);
                if(int1 <= 1){strcat(output,"000");}
                else if(int1 <= 3){strcat(output,"00");}
                else if(int1 <= 7){strcat(output,"0");}
                dec2bin(int1);
                strcat(output,dec_out);
            }
            else if(strcmp(operand,"SRH0")==0)
            {
                strcpy(output,"0101");
                fscanf(in_file,"%d\n",int1);
                printf("int1 = %d\n",int1);
                if(int1 <= 1){strcat(output,"000");}
                else if(int1 <= 3){strcat(output,"00");}
                else if(int1 <= 7){strcat(output,"0");}
                dec2bin(int1);
                strcat(output,dec_out);
            }
            else if(strcmp(operand,"CLR")==0)
            {
                strcpy(output,"011000");
                fscanf(in_file,"%s\n",str1);
                printf("str1 = %s\n",str1);
                reg2bin(str1);
                strcat(output,reg_out);
            }
            else if(strcmp(operand,"MOV")==0)
            {
                printf("operand = %s\n",operand);
                strcpy(output,"0111");
                fscanf(in_file,"%s, str2\n",str1,str2);
                printf("str1 = %s\n",str1);
                printf("str2 = %s\n",str2);
                reg2bin(str1);
                strcat(output,reg_out);
                reg2bin(str2);
                strcat(output,reg_out);
            }
            else if(strcmp(operand,"MOVA")==0)
            {
                strcpy(output,"110000");
                fscanf(in_file,"%s\n",str1);
                printf("str1 = %s\n",str1);
                reg2bin(str1);
                strcat(output,reg_out);
            }
            else if(strcmp(operand,"MOVR")==0)
            {
                strcpy(output,"110001");
                fscanf(in_file,"%s\n",str1);
                printf("str1 = %s\n",str1);
                reg2bin(str1);
                strcat(output,reg_out);
            }
            else if(strcmp(operand,"MOVRHS")==0)
            {
                strcpy(output,"110010");
                fscanf(in_file,"%s\n",str1);
                printf("str1 = %s\n",str1);
                reg2bin(str1);
                strcat(output,reg_out);
            }
            else if(strcmp(operand,"PAUSE")==0)
            {
                strcpy(output,"00000000");
            }

            counter = 0;
            strcpy(operand,"");
            fprintf(out_file,"%s\n",output);
        }
        // printf("%s == SR0 => %d\n",operand,strcmp(operand,"SR0"));

    }
    fclose(in_file);
    fclose(out_file);
}

File: Lab5_hank/tutorial/compiler.py
def num2bin(val,bits):
    if val >= 0:
        tmp = bin(val)
        return tmp[2:]
    else:
        tmp = -1*val
        tmp = bin(tmp-(1<<bits))
        return tmp[3:]
    

def reg2bin(string):
    if string == "R0":
        return("00")
    elif string == "R1":
        return("01")
    elif string == "R2":
        return("10")
    elif string == "R3":
        return("11")
    else:
        return(-1)

def line2bin(line):
    if line[0] == "BR":
        tmp = num2bin(int(line[1]),5)
        writeline = "100" + (5-len(tmp)) * "0"+tmp
    elif line[0] == "BRZ":
        tmp = num2bin(int(line[1]),5)
        writeline = "101" + (5-len(tmp))*"0" + tmp
    elif line[0] == "ADDI":
        tmp = num2bin(int(line[2]),3)
        writeline = "000" + (3-len(tmp))*"0" + tmp + reg2bin(line[1])
    elif line[0] == "SUBI":
        tmp = num2bin(int(line[2]),3)
        writeline = "001" + (3-len(tmp))*"0" + tmp + reg2bin(line[1])
    elif line[0] == "SR0":
        tmp = num2bin(int(line[1]),4)
        writeline = "0100" + (4-len(tmp))*"0" + tmp
    elif line[0] == "SRH0":
        tmp = num2bin(int(line[1]),4)
        writeline = "0101" + (4-len(tmp))*"0" + tmp
    elif line[0] == "CLR":
        writeline = "011000"+ reg2bin(line[1])
    elif line[0] == "MOV":
        writeline = "0111" + reg2bin(line[1]) + reg2bin(line[2])
    elif line[0] == "MOVA":
        writeline = "110000"+ reg2bin(line[1])
    elif line[0] == "MOVR":
        writeline = "110001"+ reg2bin(line[1])
    elif line[0] == "MOVRHS":
        writeline = "110010"+ reg2bin(line[1])
    elif line[0] == "PAUSE":
        writeline = 8*"1"
    else:
        writeline = ""
    return(writeline)

def str2line(line):
    line = line.strip("\n")
    line = line.split(" ")
    line2 = []
    for i in line:
        i = i.split(",")
        for j in i:
            if j != '':
                line2.append(j)
    if line2 != []:
        line = line2
    return(line)

def main():
    filename = input("file name:")
    infile = open(filename,"r")
    outfile = open("instruction_rom.mif","w")
    ## header of file
    outfile.write("-- Copyright (C) 2017  Intel Corporation. All rights reserved.\n")
    outfile.write("-- Your use of Intel Corporation's design tools, logic functions \n")
    outfile.write("-- and other software and tools, and its AMPP partner logic \n")
    outfile.write("-- functions, and any output files from any of the foregoing \n")
    outfile.write("-- (including device programming or simulation files), and any \n")
    outfile.write("-- associated documentation or information are expressly subject \n")
    outfile.write("-- to the terms and conditions of the Intel Program License \n")
    outfile.write("-- Subscription Agreement, the Intel Quartus Prime License Agreement,\n")
    outfile.write("-- the Intel FPGA IP License Agreement, or other applicable license\n")
    outfile.write("-- agreement, including, without limitation, that your use is for\n")
    outfile.write("-- the sole purpose of programming logic devices manufactured by\n")
    outfile.write("-- Intel and sold by Intel or its authorized distributors.  Please\n")
    outfile.write("-- refer to the applicable agreement for further details.\n\n")
    outfile.write("-- Quartus Prime generated Memory Initialization File (.mif)\n\n")
    outfile.write("WIDTH=8;\nDEPTH=256;\n\n")
    outfile.write("ADDRESS_RADIX=UNS;\nDATA_RADIX=BIN;\n\n")
    outfile.write("CONTENT BEGIN\n")
##    outfile.write("\t0    :   00000000;\n")
    # content
    line = infile.readline()
    cntr = 0
    while line != '':
        if line[0] != "#":
            line = str2line(line)
            print(line)
            writeline = line2bin(line)
            print(writeline)
            outfile.write("\t"+str(cntr)+(5-len(str(cntr)))*" "+":   "+writeline+";\n")
            cntr+=1
        line = infile.readline()
    outfile.write("\t["+str(cntr)+"..255]  :   11111111;\n")
    outfile.write("END;\n")
    outfile.close()
    infile.close()

main()
